### PR TITLE
Create rpm installer for FLASH code.

### PR DESCRIPTION
**Pull request by e-carlin**
Each FLASH example has it's own binary and setup_units file.

### CODE REVIEW

File: installers/rpm-code/codes/flash.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#!/bin/bash
+
+flash_main() {
+    codes_dependencies hypre
+    # TODO(e-carlin):  is $install_server right? rn probably already has an abstraction for this
+    # TODO(e-carlin):  need to make this url secret with the secret coming from the env
+    codes_download "$install_server/flash-4.6.2.tar.gz" "FLASH4.6.2"
```
Comment (e-carlin) [inline installers/rpm-code/codes/flash.sh:7]: @robnagler is there an abstraction already for this? I couldn't find one... should one be created?

How should we share the secret part of the url? Also should the code live at $install_server/flash? Do we need a dev/alpha/beta part of the path?

File: installers/rpm-code/codes/flash.sh [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+157c157
+< #----------------------------------------------------------------------------
+---
+> #----------------------------------------------------------------------------
+EOF
+}
+
+setup_CapLaser() {
+    cd source/Simulation/SimulationMain/magnetoHD
+    # TODO(e-carlin):  is $install_server right? I would guess rn already has an abstraction for this
```
Comment (e-carlin) [inline installers/rpm-code/codes/flash.sh:47]: [Same question as above](https://github.com/radiasoft/download/pull/95/files#r421828850)

File: installers/rpm-code/codes/flash.sh (hunk 1 above)
Comment (robnagler) [reply installers/rpm-code/codes/flash.sh:7]: With #97 this will be: 

```py
codes_download_proprietary "$install_server/flash-4.6.2.tar.gz" "FLASH4.6.2"
```

`install_proprietary_key` is the magic value, and it is cascaded by `install_vars_export` for `vagrant ssh` and `docker run` used by build scripts.

Comment (robnagler) [reply installers/rpm-code/codes/flash.sh:7]: For the build, we do not need an alpha/beta/etc. That's promotion, and handled by our distro mechanisms and rsconf.

**Comment by e-carlin** [comment]
@robnagler now using codes_download_proprietary https://github.com/radiasoft/download/pull/95/commits/7ce2e4e0347077604db225c9143346c6cb2aebcf. Read for another review.

File: installers/rpm-code/codes/flash.sh [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    install -m 444 setup_units "${codes_dir[share]}"/flash4/setup_units-$1
+    cd ..
+}
+
+patch_makefile() {
+    patch sites/Prototypes/Linux/Makefile.h <<'EOF'
+6c6
+< MPI_PATH   = /usr/local/mpich2/
+---
+> MPI_PATH   = /usr/lib64/mpich/
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:29]:
```sh
+ > MPI_PATH   = $d/
```

File: installers/rpm-code/codes/flash.sh [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+8,9c8,9
+< HDF5_PATH  = /usr/local/hdf5
+< HYPRE_PATH = /usr/local/hypre
+---
+> HDF5_PATH  = /usr/lib64/mpich
+> HYPRE_PATH = /home/vagrant/.local
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:35]:
```sh
+ > HDF5_PATH  = $d
+ > HYPRE_PATH = ${codes_dir[prefix]}
```

File: installers/rpm-code/codes/flash.sh [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+< #----------------------------------------------------------------------------
+---
+> #----------------------------------------------------------------------------
+EOF
+}
+
+setup_CapLaser() {
+    cd source/Simulation/SimulationMain/magnetoHD
+    codes_download_proprietary "flash/CapLaserBELLA-4.6.2.tar.gz" "CapLaserBELLA"
+    cd $flash_src_root
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:46]: Don't use a global unless you have to. One way to handle this is to surround the whole thing in `()`, but this `cd` shouldn't be necessary.

File: installers/rpm-code/codes/flash.sh [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+> HYPRE_PATH = /home/vagrant/.local
+157c157
+< #----------------------------------------------------------------------------
+---
+> #----------------------------------------------------------------------------
+EOF
+}
+
+setup_CapLaser() {
+    cd source/Simulation/SimulationMain/magnetoHD
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:44]: This `cd` is unecessary. Instead you should embed this path in the CapLaserBELLA tarball. There's no need for this code to know about `source/Simulation/SimulationMain/magnetoHD`.

File: installers/rpm-code/codes/flash.sh [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    cd $flash_src_root
+    ./setup -auto magnetoHD/CapLaser -2d -nxb=16 -nyb=16 +hdf5typeio \
+            species=fill,wall +mtmmmt +usm3t +mgd mgd_meshgroups=6 \
+            -parfile=caplaser_basic.par -objdir=$1 +laser \
+            ed_maxPulses=1 ed_maxPulseSections=4 ed_maxBeams=1
+
+}
+
+setup_RTFlame() {
+    ./setup RTFlame -2d -auto -nxb=16 -nyb=16 -objdir=$1
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:55]: The `$1` and and RTFlame should be one and the same. We want flashType to be the same as the `objdir` so that we don't get collisions on names (you will be compiling multiple codes within the same directory structure).

File: installers/rpm-code/codes/flash.sh [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+---
+> #----------------------------------------------------------------------------
+EOF
+}
+
+setup_CapLaser() {
+    cd source/Simulation/SimulationMain/magnetoHD
+    codes_download_proprietary "flash/CapLaserBELLA-4.6.2.tar.gz" "CapLaserBELLA"
+    cd $flash_src_root
+    ./setup -auto magnetoHD/CapLaser -2d -nxb=16 -nyb=16 +hdf5typeio \
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:47]: Does CapLaser need to be magnetoHD? What happens if you make it exactly like RTFlame, and put it one level up? Are there includes in the CaplaserBELLA that are relative to the directory (or are they using `cc -I` paths).

File: installers/rpm-code/codes/flash.sh [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+< MPI_PATH   = /usr/local/mpich2/
+---
+> MPI_PATH   = /usr/lib64/mpich/
+8,9c8,9
+< HDF5_PATH  = /usr/local/hdf5
+< HYPRE_PATH = /usr/local/hypre
+---
+> HDF5_PATH  = /usr/lib64/mpich
+> HYPRE_PATH = /home/vagrant/.local
+157c157
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:36]: remove this

File: installers/rpm-code/codes/flash.sh [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    do
+        setup_$n $n
+        make_and_install $n
+    done
+}
+
+make_and_install() {
+    cd $1
+    codes_make
+    install -m 755 flash4 "${codes_dir[bin]}"/flash4-$1
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:18]: Quote `"$1"`. Document the `flash4-$1` with a POSIT so we know what has to match.

File: installers/rpm-code/codes/flash.sh [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    install -d 755 "${codes_dir[share]}"/flash4
+    install -m 444 setup_units "${codes_dir[share]}"/flash4/setup_units-$1
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:20]: Quote `"$1"`. Document `flash4/setup_units-$1` with a POSIT.

File: installers/rpm-code/codes/flash.sh [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        setup_$n $n
+        make_and_install $n
+    done
+}
+
+make_and_install() {
+    cd $1
+    codes_make
+    install -m 755 flash4 "${codes_dir[bin]}"/flash4-$1
+    install -d 755 "${codes_dir[share]}"/flash4
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:19]: This only needs to be done once and maybe could then be passed in to `make_and_install`.

File: installers/rpm-code/codes/flash.sh [hunk 13]
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+
+flash_main() {
+    codes_dependencies hypre
+    codes_download_proprietary "flash/FLASH-4.6.2.tar.gz" "FLASH4.6.2"
+    local flash_src_root=$PWD
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:6]: remove (see below).

File: installers/rpm-code/codes/flash.sh [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    patch_makefile
+    for n in CapLaser RTFlame
+    do
+        setup_$n $n
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:10]: Embed `setup_$n $n` inside make_and_install, since it's not really relevant here to know that setup exists. Quote `"setup_$n" "$n"`.

All variables in bash should be quoted except in cases where they shouldn't :)

Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 13 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 12 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2****

File: installers/rpm-code/codes/flash.sh (hunk 11 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 10 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 6 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 7 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 8 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 9 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 5 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 4 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh (hunk 3 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    cd $1
+    codes_make
+    install -m 755 flash4 "${codes_dir[bin]}"/flash4-$1
+    install -d 755 "${codes_dir[share]}"/flash4
+    install -m 444 setup_units "${codes_dir[share]}"/flash4/setup_units-$1
+    cd ..
+}
+
+patch_makefile() {
+    patch sites/Prototypes/Linux/Makefile.h <<'EOF'
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

File: installers/rpm-code/codes/flash.sh [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    local flash_src_root=$PWD
+    patch_makefile
+    for n in CapLaser RTFlame
+    do
+        setup_$n $n
+        make_and_install $n
+    done
+}
+
+make_and_install() {
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4b4294d531035f961f3421d9e7450d2b8d2c07b2

**Comment by e-carlin** [comment]
@robnagler all comments addressed.

File: installers/rpm-code/codes/flash.sh [hunk 17]
```diff
@@ -3,54 +3,51 @@
...
-    cd $flash_src_root
-    ./setup -auto magnetoHD/CapLaser -2d -nxb=16 -nyb=16 +hdf5typeio \
+flash_setup_CapLaserBELLA() {
+    codes_download_proprietary "flash/$1-4.6.2.tar.gz" "source"
+    cd ..
+    ./setup "$1" -objdir="$1" -auto -2d -nxb=16 -nyb=16 +hdf5typeio \
             species=fill,wall +mtmmmt +usm3t +mgd mgd_meshgroups=6 \
-            -parfile=caplaser_basic.par -objdir=$1 +laser \
-            ed_maxPulses=1 ed_maxPulseSections=4 ed_maxBeams=1
+            -parfile=caplaser_basic.par +laser \ ed_maxPulses=1
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:46]: this line seems incorrect. The next line only works, because it is correct shell syntax

File: installers/rpm-code/codes/flash.sh [hunk 18]
```diff
@@ -3,54 +3,51 @@
...
-make_and_install() {
-    cd $1
+flash_make_and_install_type() {
+    "flash_setup_$n" "$n"
+    cd "$1"
     codes_make
-    install -m 755 flash4 "${codes_dir[bin]}"/flash4-$1
-    install -d 755 "${codes_dir[share]}"/flash4
-    install -m 444 setup_units "${codes_dir[share]}"/flash4/setup_units-$1
+    # POSIT: Sirepo assumes the exe is named flash4-flashType
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:18]: `# POSIT: sirepo.sim_data.flash.SimData.flash_exe_path assumes flash4-flashType`

File: installers/rpm-code/codes/flash.sh [hunk 19]
```diff
@@ -3,54 +3,51 @@
...
+    install -m 755 flash4 "${codes_dir[bin]}"/flash4-"$1"
+    # POSIT: Sirepo assumes the setup_units file is named setup_units-flashType
```
Comment (robnagler) [inline installers/rpm-code/codes/flash.sh:20]: `# POSIT: sirepo.sim_data.flash.SimData.flash_setup_units_path assumes flash4/setup_units-flashType`

File: installers/rpm-code/codes/flash.sh (hunk 18 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: I'll get the right verbosity one of these days... :)

https://github.com/radiasoft/download/pull/95/commits/4f394e009f65fb6649b1f4a0d9103ddd9be63bde

File: installers/rpm-code/codes/flash.sh (hunk 19 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4f394e009f65fb6649b1f4a0d9103ddd9be63bde

File: installers/rpm-code/codes/flash.sh (hunk 17 above)
Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: Good catch. Emacs fill-region doesn't work here. I guess it probably can't. Too hard to parse.

Comment (e-carlin) [reply installers/rpm-code/codes/flash.sh]: https://github.com/radiasoft/download/pull/95/commits/4f394e009f65fb6649b1f4a0d9103ddd9be63bde
